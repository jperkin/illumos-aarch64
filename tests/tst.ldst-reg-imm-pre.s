/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.9 Load/store register (immediate pre-indexed)
 *
 * C5.6.180 STRB  <Wt>, [<Xn|SP>, #<simm>]!
 * C5.6.86  LDRB  <Wt>, [<Xn|SP>, #<simm>]!
 * C5.6.90  LDRSB <Rt>, [<Xn|SP>, #<simm>]!
 * C6.3.285 STR   <Vt>, [<Xn|SP>, #<simm>]!
 * C6.3.166 LDR   <Vt>, [<Xn|SP>, #<simm>]!
 * C5.6.182 STRH  <Wt>, [<Xn|SP>, #<simm>]!
 * C5.6.88  LDRH  <Wt>, [<Xn|SP>, #<simm>]!
 * C5.6.92  LDRSH <Rt>, [<Xn|SP>, #<simm>]!
 * C5.6.178 STR   <Rt>, [<Xn|SP>, #<simm>]!
 * C5.6.83  LDR   <Rt>, [<Xn|SP>, #<simm>]!
 * C5.6.94  LDRSW <Xt>, [<Xn|SP>, #<simm>]!
 *
 * simm is signed 9bit, so test min/max.
 */
.macro ldst_reg_imm_pre op, wreg=0, xreg=0, simd=0
	.if \wreg
	\op	w0,  [x0,  #0]!
	\op	w30, [x30, #255]!
	\op	wzr, [sp,  #-256]!
	.endif
	.if \xreg
	\op	x0,  [x0,  #0]!
	\op	x30, [x30, #255]!
	\op	xzr, [sp,  #-256]!
	.endif
	.if \simd
	\op	b0,  [x0, #0]!
	\op	h31, [sp, #1]!
	\op	s0,  [x0, #-1]!
	\op	d31, [sp, #255]!
	\op	q0,  [x0, #-256]!
	.endif
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldst_reg_imm_pre strb  1
	ldst_reg_imm_pre ldrb  1
	ldst_reg_imm_pre ldrsb 1 1
	ldst_reg_imm_pre str   1 1 1
	ldst_reg_imm_pre ldr   1 1 1
	ldst_reg_imm_pre strh  1
	ldst_reg_imm_pre ldrh  1
	ldst_reg_imm_pre ldrsh 1 1
	ldst_reg_imm_pre ldrsw 0 1
.size libdis_test, [.-libdis_test]
