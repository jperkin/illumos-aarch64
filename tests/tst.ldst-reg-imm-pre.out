    libdis_test:      00 0c 00 38  strb w0, [x0, #0]!
    libdis_test+0x4:  de ff 0f 38  strb w30, [x30, #255]!
    libdis_test+0x8:  ff 0f 10 38  strb wzr, [sp, #-256]!
    libdis_test+0xc:  00 0c 40 38  ldrb w0, [x0, #0]!
    libdis_test+0x10: de ff 4f 38  ldrb w30, [x30, #255]!
    libdis_test+0x14: ff 0f 50 38  ldrb wzr, [sp, #-256]!
    libdis_test+0x18: 00 0c c0 38  ldrsb w0, [x0, #0]!
    libdis_test+0x1c: de ff cf 38  ldrsb w30, [x30, #255]!
    libdis_test+0x20: ff 0f d0 38  ldrsb wzr, [sp, #-256]!
    libdis_test+0x24: 00 0c 80 38  ldrsb x0, [x0, #0]!
    libdis_test+0x28: de ff 8f 38  ldrsb x30, [x30, #255]!
    libdis_test+0x2c: ff 0f 90 38  ldrsb xzr, [sp, #-256]!
    libdis_test+0x30: 00 0c 00 b8  str w0, [x0, #0]!
    libdis_test+0x34: de ff 0f b8  str w30, [x30, #255]!
    libdis_test+0x38: ff 0f 10 b8  str wzr, [sp, #-256]!
    libdis_test+0x3c: 00 0c 00 f8  str x0, [x0, #0]!
    libdis_test+0x40: de ff 0f f8  str x30, [x30, #255]!
    libdis_test+0x44: ff 0f 10 f8  str xzr, [sp, #-256]!
    libdis_test+0x48: 00 0c 00 3c  str b0, [x0, #0]!
    libdis_test+0x4c: ff 1f 00 7c  str h31, [sp, #1]!
    libdis_test+0x50: 00 fc 1f bc  str s0, [x0, #-1]!
    libdis_test+0x54: ff ff 0f fc  str d31, [sp, #255]!
    libdis_test+0x58: 00 0c 90 3c  str q0, [x0, #-256]!
    libdis_test+0x5c: 00 0c 40 b8  ldr w0, [x0, #0]!
    libdis_test+0x60: de ff 4f b8  ldr w30, [x30, #255]!
    libdis_test+0x64: ff 0f 50 b8  ldr wzr, [sp, #-256]!
    libdis_test+0x68: 00 0c 40 f8  ldr x0, [x0, #0]!
    libdis_test+0x6c: de ff 4f f8  ldr x30, [x30, #255]!
    libdis_test+0x70: ff 0f 50 f8  ldr xzr, [sp, #-256]!
    libdis_test+0x74: 00 0c 40 3c  ldr b0, [x0, #0]!
    libdis_test+0x78: ff 1f 40 7c  ldr h31, [sp, #1]!
    libdis_test+0x7c: 00 fc 5f bc  ldr s0, [x0, #-1]!
    libdis_test+0x80: ff ff 4f fc  ldr d31, [sp, #255]!
    libdis_test+0x84: 00 0c d0 3c  ldr q0, [x0, #-256]!
    libdis_test+0x88: 00 0c 00 78  strh w0, [x0, #0]!
    libdis_test+0x8c: de ff 0f 78  strh w30, [x30, #255]!
    libdis_test+0x90: ff 0f 10 78  strh wzr, [sp, #-256]!
    libdis_test+0x94: 00 0c 40 78  ldrh w0, [x0, #0]!
    libdis_test+0x98: de ff 4f 78  ldrh w30, [x30, #255]!
    libdis_test+0x9c: ff 0f 50 78  ldrh wzr, [sp, #-256]!
    libdis_test+0xa0: 00 0c c0 78  ldrsh w0, [x0, #0]!
    libdis_test+0xa4: de ff cf 78  ldrsh w30, [x30, #255]!
    libdis_test+0xa8: ff 0f d0 78  ldrsh wzr, [sp, #-256]!
    libdis_test+0xac: 00 0c 80 78  ldrsh x0, [x0, #0]!
    libdis_test+0xb0: de ff 8f 78  ldrsh x30, [x30, #255]!
    libdis_test+0xb4: ff 0f 90 78  ldrsh xzr, [sp, #-256]!
    libdis_test+0xb8: 00 0c 80 b8  ldrsw x0, [x0, #0]!
    libdis_test+0xbc: de ff 8f b8  ldrsw x30, [x30, #255]!
    libdis_test+0xc0: ff 0f 90 b8  ldrsw xzr, [sp, #-256]!
